packages/flutter_tools/lib/src/build_system/targets/common.dart

```diff
@@ -281,6 +287,41 @@ class KernelSnapshot extends Target {
       throw Exception();
     }
   }
+
+  Future<void> _addFlavorToDartDefines(
+    List<String> dartDefines,
+    Environment environment,
+    TargetPlatform targetPlatform,
+  ) async {
+    final String? flavor;
+
+    // For iOS and macOS projects, parse the flavor from the configuration, do
+    // not get from the FLAVOR environment variable. This is because when built
+    // from Xcode, the scheme/flavor can be changed through the UI and is not
+    // reflected in the environment variable.
+    if (targetPlatform == TargetPlatform.ios || targetPlatform == TargetPlatform.darwin) {
+      final FlutterProject flutterProject = FlutterProject.fromDirectory(environment.projectDir);
+      final XcodeBasedProject xcodeProject =
+          targetPlatform == TargetPlatform.ios ? flutterProject.ios : flutterProject.macos;
+      flavor = await xcodeProject.parseFlavorFromConfiguration(environment);
+    } else {
+      flavor = environment.defines[kFlavor];
+    }
+    if (flavor != null) {
+      if (globals.platform.environment['FLUTTER_APP_FLAVOR'] != null) {
```

Comment:
nit: Consider moving this into `kAppFlavor` or similar since we reference it everywhere.

This can be a in a follow-up PR, since it would be nice to also change existing calls.